ast/src/main/java/com/graphicsfuzz/common/ast/visitors/AstBuilder.java

```diff
@@ -195,9 +198,124 @@ void addTopLevelDeclaration(Declaration decl) {
+          }
+
+          int lval = ((IntConstantExpr) lexpr).getNumericValue();
+          int rval = ((IntConstantExpr) rexpr).getNumericValue();
+          int fval = 0;
+          switch (bexpr.getOp()) {
+            case MOD:
+              fval = lval % rval;
+              break;
+            case MUL:
+              fval = lval * rval;
+              break;
+            case DIV:
+              if (rval == 0) {
+                throw new RuntimeException("Division by zero folding constant " + bexpr.getText());
+              }
+              fval = lval / rval;
+              break;
+            case ADD:
+              fval = lval + rval;
+              break;
+            case SUB:
+              fval = lval - rval;
+              break;
+            case BAND:
+              fval = lval & rval;
+              break;
+            case BOR:
+              fval = lval | rval;
+              break;
+            case BXOR:
+              fval = lval ^ rval;
+              break;
+            case LAND:
+              fval = ((lval != 0) && (rval != 0)) ? 1 : 0;
+              break;
+            case LOR:
+              fval = ((lval != 0) || (rval != 0)) ? 1 : 0;
+              break;
+            case LXOR:
+              fval = ((lval == 0) != (rval == 0)) ? 1 : 0;
+              break;
+            case SHL:
+              fval = lval << rval;
+              break;
+            case SHR:
+              fval = lval >> rval;
+              break;
+            case LT:
+              fval = (lval < rval) ? 1 : 0;
+              break;
+            case GT:
+              fval = (lval > rval) ? 1 : 0;
+              break;
+            case LE:
+              fval = (lval <= rval) ? 1 : 0;
+              break;
+            case GE:
+              fval = (lval >= rval) ? 1 : 0;
+              break;
+            case EQ:
+              fval = (lval == rval) ? 1 : 0;
+              break;
+            case NE:
+              fval = (lval != rval) ? 1 : 0;
+              break;
+            default:
+              throw new RuntimeException("Unable to fold constant " + bexpr.getText());
+          }
+          return new IntConstantExpr(Integer.toString(fval));
+        }
+        throw new RuntimeException("Unable to fold constant " + expr.getText());
+      }
+
+      @Override
+      public void visitVariableDeclInfo(VariableDeclInfo variableDeclInfo) {
+        if (variableDeclInfo.hasArrayInfo() && variableDeclInfo.getArrayInfo().hasSize()) {
+          Expr expr = variableDeclInfo.getArrayInfo().getSizeExpr();
+          expr = reduce(expr);
+          variableDeclInfo.getArrayInfo().setSizeExpr(expr);
+        }
+        super.visitVariableDeclInfo(variableDeclInfo);
+      }
+    }.visit(tu);
+    return tu;
+  }
+
   public static TranslationUnit getTranslationUnit(Translation_unitContext ctx,
                                                    ShaderKind shaderKind, boolean hasWebGlHint) {
```

Comment:
Can you add a brief comment here to explain why we fix up array sizes?  (Similar to the comment inside 'fixUpArraySizes' but briefer.)